src/com/redhat/ThreescaleService.groovy

```diff
@@ -2,72 +2,91 @@
+        def globalOptions = toolbox.getGlobalToolboxOptions()
+        def commandLine = "3scale import openapi ${globalOptions} -t ${this.environment.targetSystemName} -d ${this.toolbox.destination} /artifacts/${baseName}"
+        if (this.environment.stagingPublicBaseURL != null) {
+            commandLine += " --staging-public-base-url=${this.environment.stagingPublicBaseURL}"
+        }
+        if (this.environment.productionPublicBaseURL != null) {
+            commandLine += " --production-public-base-url=${this.environment.productionPublicBaseURL}"
+        }
+        if (this.environment.privateBaseUrl != null) {
+            commandLine += " --override-private-base-url=${this.environment.privateBaseUrl}"
+        }
+
+        toolbox.runToolbox(commandLine: commandLine,
+                jobName: "import",
+                openAPI: [
+                        "filename": baseName,
+                        "content" : util.readFile(this.openapi.filename)
+                ])
+    }
+
+    void applyApplicationPlans() {
+        def globalOptions = toolbox.getGlobalToolboxOptions()
+        this.applicationPlans.each {
+            // the issue THREESCALE-2816 for resolving update error
+            def commandLine = "3scale application-plan apply ${globalOptions} ${this.toolbox.destination} ${this.environment.targetSystemName} ${it.systemName} --approval-required=${it.approvalRequired} --cost-per-month=${it.costPerMonth} --end-user-required=${it.endUserRequired} --name=${it.name} --publish=${it.published} --setup-fee=${it.setupFee} --trial-period-days=${it.trialPeriodDays}"
+            if (it.defaultPlan) {
+                commandLine += " --default"
+            }
 
-  void importOpenAPI() {
-    Util util = new Util()
+            toolbox.runToolbox(commandLine: commandLine,
+                    jobName: "apply-application-plan-${it.systemName}")
+        }
+    }
+
+    void applyApplication(Map applicationMap) {
+
+        def globalOptions = toolbox.getGlobalToolboxOptions()
+        this.applications.each {
+            def commandLine = "3scale application apply --account=${it.account}  --plan=${it.plan}  --service=${this.environment.targetSystemName} --name=${it.name}  --description=${it.description}  ${globalOptions} ${this.toolbox.destination}   ${it.name} "
+
+            toolbox.runToolbox(commandLine: commandLine,
+                    jobName: "apply-application-${it.name}")
+        }
+    }
+
+    Map readProxy(String environment) {
 
-    if (this.environment.stagingPublicBaseURL != null || this.environment.productionPublicBaseURL != null) {
-      // See https://issues.jboss.org/browse/THREESCALE-2607
-      throw new Exception("NOT_IMPLEMENTED")
+        def globalOptions = toolbox.getGlobalToolboxOptions()
+        def commandLine = "3scale proxy-config show ${globalOptions} ${this.toolbox.destination} ${this.environment.targetSystemName} ${environment}"
+        toolbox.runToolbox(commandLine: commandLine,
```

Comment:
We need to return the proxy definition so that the Jenkins pipeline can know the actual public staging/prod base url.

Reply:
Something like:
```groovy
    def proxyDefinition = toolbox.runToolbox(commandLine: commandLine,
                                             jobName: "show-proxy")

    return new Util().readJSON(proxyDefinition.stdout).content.proxy as Map
```

Reply:
that makes sense and return proxy definition as per comments